src/routing_node.rs

```diff
@@ -836,6 +863,114 @@ impl RoutingNode {
+        match request {
+            ExternalRequest::Get(data_request) => {
+                match data_request {
+                    DataRequest::StructuredData(data_name, value) => {
+                        match self.structured_data_cache.get(&(data_name, value)) {
+                            Some(data) => Some(Data::StructuredData(data.clone())),
+                            _ => return None,
+                        }
+                    },
+                    DataRequest::ImmutableData(data_name, _) => {
+                        match self.immutable_data_cache.get(&data_name) {
+                            Some(data) => Some(Data::ImmutableData(data.clone())),
+                            _ => None,
+                        }
+                    },
+                    DataRequest::PlainData(data_name) => {
+                        match self.plain_data_cache.get(&data_name) {
+                            Some(data) => Some(Data::PlainData(data.clone())),
+                            _ => None,
+                        }
+                    }
+                }
+            },
+            _ => None
+        }
+    }
+
+    fn handle_cache_put(&mut self, response : ExternalResponse) {
+        match response {
+            ExternalResponse::Get(data, _) => {
+                match data {
+                    Data::ImmutableData(ref immutable_data) => {
+                        self.immutable_data_cache.add(immutable_data.name(),
+                                                      immutable_data.clone());
+                    },
+                    Data::PlainData(ref plain_data) => {
+                        self.plain_data_cache.add(plain_data.name(), plain_data.clone());
+                    },
+                    Data::StructuredData(ref structured_data) => {
+                        self.structured_data_cache.add(
+                            (structured_data.name(), structured_data.get_version()),
+                            structured_data.clone());
+                    },
+                }
+            },
+            _ => {}
+        };
+    }
 }
 
 fn ignore<R,E>(_result: Result<R,E>) {}
+
+fn set_lru_cache<K, V>(lru_cache: &mut LruCache<K, V>, time_to_live: Option<Duration>,
+                       capacity: Option<usize>) where K: PartialOrd + Ord + Clone, V: Clone {
+     match (time_to_live, capacity) {
+         (Some(lru_time_to_live), Some(lru_capacity)) => {
+             *lru_cache =
+                LruCache::<K, V>::with_expiry_duration_and_capacity(lru_time_to_live,
+                                                                    lru_capacity);
+         },
+         (Some(lru_time_to_live), None) => {
+             *lru_cache = LruCache::<K, V>::with_expiry_duration(lru_time_to_live);
+         },
+         (None, Some(lru_capacity)) => {
+             *lru_cache = LruCache::<K, V>::with_capacity(lru_capacity);
+         },
+         (None, None) => {}
```

Comment:
This is not the expected behaviour; as setting none, none, now leaves it unchanged, where it would be expected to be changed to not having the cache; this function leaves the existing cache.  Rather than having this templated construction, an option on the actual cache seems simpler and more correct, would it not?

If any templating should occur, why not template it on the actual data types that are there?